File: Site Mirror/www.cas.mcmaster.ca/_se2s03/slides/OO/Song.scala
/*
 * Filename: Song.scala
 * Description:
 * -> This class DESCRIBES songs.
 * --> Funnily enough it's also an example
 * ---> of a CLASS DECLARATION
 * Author: 2S03MakeYouLearnGood
 */

// Our constructor is up here!!
class Song(n:String, a:String, r:Int) {
  require( r != 0 ) // this is the equivalent of isRuntimeValid()

  /* ----- PROPERTIES ----- */

  /*
   * FIELD VARIABLES
   *
   * They tell you the PROPERTIES of
   * the class, in this case we're
   * saying that a song has a:
   * name, type of String
   * runtime, type of int
   */
  private var name:String = n
  private var artist:String = a
  private var runtime:Int = r

  /*
   * INSTANCE METHODS
   * These are going to continue
   * the description of the behaviour
   * of a song.
   *
   * We're going to have 2 types of
   * instance methods ACCESSORS
   * and MUTATORS
   *
   * We want to be able to change and
   * retrieve various things.
   * i.e. ACCESS and MUTATE
   * e.g both ACCESS AND MUTATE
   * the runtime, name, or artist
   */
  def getName(): String = name
  def getArtist(): String = artist
  def getRuntime(): Int = runtime

  def setName(newName: String) { name = newName }

  def setArtist(newArtist: String) { artist = newArtist }

  def setRuntime(newRuntime: Int) { runtime = newRuntime }

  override def toString(): String =
    "Song: " + name + "\n" +
    "By: " + artist + "\n" +
    "Runtime: " + runtime + " seconds"

}

File: Site Mirror/www.cas.mcmaster.ca/_se2s03/slides/Class/Song.java
/*
 * Filename: Song.java
 * Description:
 * -> This class DESCRIBES songs.
 * --> Funnily enough it's also an example
 * ---> of a CLASS DECLARATION
 * ----> Now! It also keeps track of how
 * -----> many rap or rock songs!!
 * Author: 2S03MakeYouLearnGood
 */

public class Song {

  /* ----- PROPERTIES ----- */

  /* (+) FIELD VARIABLES */

  private String name;
  private String artist;
  private int runtime;
  private int genre;

  /* (-) FIELD VARIABLES */


  /* (+) STATIC VARIABLES */

  public static int numOfSongs = 0;

  /* (-) STATIC VARIABLES */


  /* (+) CONSTRUCTORS */
  // They have NO RETURN TYPE, see?

  public Song() {
    this.name = "Name of Song";
    this.artist = "Song Artist";
    this.runtime = 0;
    this.genre = 0;

    updateNumberOfSongs();
  }

  // see?
  public Song(String name, String artist) {
    this.name = name;
    this.artist = artist;
    this.runtime = 0;
    this.genre = 0;

    updateNumberOfSongs();
  }

  // see?
  public Song(String name, String artist, int runtime) {
    this.name = name;
    this.artist = artist;
    this.runtime = runtime;
    this.genre = 0;

    updateNumberOfSongs();
  }

  // see?
  public Song(String name, String artist, int runtime, int genre) {
    this.name = name;
    this.artist = artist;
    this.runtime = runtime;
    this.genre = genre;

    updateNumberOfSongs();
  }

  /* (-) CONSTRUCTORS */


  /* (+) INSTANCE METHODS */
  public String getName() { return this.name; }
  public String getArtist() { return this.artist; }
  public int getRuntime() { return this.runtime; }
  public int getGenre() { return this.genre; }

  public void setName(String name) { this.name = name; }
  public void setArtist(String artist) { this.artist = artist; }
  public void setRuntime(int runtime) {
    if( runtime < 0 )
      throw new IllegalArgumentException("That runtime is ****ed up.");

    this.runtime = runtime;
  }
  public void setGenre(int genre) {
    if( genre != 0 && genre != 1)
      throw new IllegalArgumentException("'0' or '1' please!");
    this.genre = genre;
  }

  public String toString() {
    String g = "";
    switch (this.genre) {
      case 0: g = "Rap"; break;
      case 1: g = "Rock"; break;
    }
    return "Song: " + this.name + "\n" +
           "By: " + this.artist + "\n" +
           "Runtime: " + this.runtime + " seconds\n" +
           "Genre: " + g + "\n";
  }

  /* (-) INSTANCE METHODS */


  /* (+) STATIC METHODS */

  // Before we make a METHOD (nay! anything)
  // PUBLIC STATIC
  // we need to ask ourselves the question:
  // -> Does it make sense to
  // --> CALL this METHOD (or otherwise)
  // ---> if we've not created ANY objects?
  // ----> this is a STOOPID PUBLIC STATIC METHOD
  public static String creator() {
    return "2S03MakeYouLearnGood";
  }

  private static void updateNumberOfSongs() {
    numOfSongs++;
  }

  /* (-) STATIC METHODS */
}